crates/editor/src/element.rs

```diff
@@ -3683,24 +3684,98 @@ impl LineWithInvisibles {
+                if !selection_ranges
                     .iter()
                     .any(|region| region.start <= invisible_point && invisible_point < region.end)
                 {
-                    continue;
+                    return;
+                }
+
+                paint(cx);
+            }),
+
+            // For a whitespace to be on a boundary, any of the following conditions need to be met:
+            // - It is a tab
+            // - It is adjacent to an edge (start or end)
+            // - It is adjacent to a whitespace (left or right)
+            ShowWhitespaceSetting::Boundary => {
+                // We'll need to keep track of the last invisible we've seen and then check if we are adjacent to it for some of
+                // the above cases.
+                // Note: We zip in the original `invisibles` to check for tab equality
+                let mut last_seen: Option<(bool, usize, Box<dyn Fn(&mut WindowContext)>)> = None;
+                for (([start, end], paint), invisible) in
+                    invisible_iter.zip_eq(self.invisibles.iter())
+                {
+                    let should_render = match (&last_seen, invisible) {
+                        // Short out early if the invisible is a tab character
+                        (_, Invisible::Tab { .. }) => true,
+
+                        // Otherwise, we need to see if the last invisible is adjacent to us
+                        (Some((_, last_end, _)), _) => *last_end == start,
+
+                        // All other cases are false
+                        _ => false,
+                    };
+
+                    // Render the whitespace if the above checks are valid, also making sure to always render starting and ending whitespace
+                    if start == 0 || end == self.line.len || should_render {
+                        paint(cx);
+
+                        // Since we are scanning from the left, we will skip over the first available whitespace that is part
+                        // of a boundary between non-whitespace segments, so we correct by manually redrawing it if needed.
+                        if let Some((should_render_last, last_end, paint_last)) = last_seen {
+                            // Note that we need to make sure that the last one is actually adjacent
+                            if !should_render_last && last_end == start {
+                                paint_last(cx);
+                            }
+                        }
+                    }
+
+                    // Manually render anything within a selection
+                    let invisible_point = DisplayPoint::new(row, start as u32);
+                    if selection_ranges.iter().any(|region| {
+                        region.start <= invisible_point && invisible_point < region.end
+                    }) {
+                        paint(cx);
+                    }
+
+                    // Keep track of the last for the next iteration
```

Comment:
To note, most of the comments that explain what the code does, such as 

> Helper to transform each Invisible into parts needed to optionally paint it later

> Start rendering the whitespace

> Short out early if the invisible is a tab character

> Otherwise, we need to see if the last invisible is adjacent to us

> Keep track of the last for the next iteration

> All other cases are false

> Render the whitespace if the above checks are valid, also making sure to always render starting and ending whitespace


etc.

are useless, as the code itself shows that pretty well.
I would remove them all and only leave the ones that describe the "external" knowledge, not seen in the code directly (such as the adjacent whitespace description) and some caveats ("Since we are scanning from the left, " thing seems a bit like it).

Up to you though.

Reply:
Removed!